README
/*
 * Argus Software - README
 * Copyright (c) 2000-2012 QoSient, LLC
 * All rights reserved. 
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * 
 */


Argus 3.0.6
QoSient, LLC
http://qosient.com/argus

This directory contains source code for Argus, an IP network
transaction auditing tool, that generates a persistent audit of all
network transactions and their performance.  The data that Argus
generates can be used for a wide range of tasks including
Network Operations, Security, Assurance and Performance Management.

To install the software and get started, see the ./INSTALL file.
For answers to many questions, please see the ./docs/FAQ file.

The Argus 3.0 has been ported to SunOS and Solaris, Linux, FreeBSD,
OpenBSD, NetBSD and Irix, and the sample client code has
been successfully used under most variants of Unix.  If you port
Argus to another system, please let us know on the mailing list.

Problems, bugs, questions, desirable enhancements, source code
contributions, etc., should be sent to argus-info@lists.andrew.cmu.edu

************************************************************

IMPORTANT:

Send problems, bug reports, questions and comments to the argus
mailing list at Carnegie Mellon University:

argus-info@lists.andrew.cmu.edu.

Please use the './bin/argusbug' script when posting bug reports
about Argus. Argusbug will gather some information about your
system and start your editor with a form in which you can describe
your problem.  Delete information that you consider non-relevant
to your problem.  Bug reports not generated by Argusbug might
be silently ignored by the Argus maintainers, so please use the
tool.  A report that says 'Argus does not work. Why?' will
not be considered a valid bug report.

************************************************************

Comprehensive network transaction auditing is an extremely powerful
network management tool, and a large number of sites can benefit from
the prototype work that has been done through the Argus Project.

Again, thank you for your interest in Argus.  I hope that you find
the software useful.


Carter Bullard
carter@qosient.com



ChangeLog
Tue Apr 10 08:19:44 EDT 2012
Prepare for argus-3.0.6 release.
Modify usage to use stdout, instead of stderr

Mon Feb 13 11:11:32 EST 2012
Fixed embedded comments parsing in argus.conf file.
Updated argus client filter to be in sync with clients.
Fixed issue with ArgusSourceID parsing.

Wed Feb  1 22:25:37 EST 2012
Fixed loss reporting bug.  Adjusted TCP PERF DSR to support
gap estimates better.  Need to modify for argus-3.1.0 so
that we track retransmitted bytes, rather than packets (possibly).

Tue Jan  3 18:47:20 EST 2012
Added better double quote handling in the argus resource file.
Removed all options in the sample argus.conf file.
Preliminary Juniper encapsulation support added

Tue Dec  6 20:19:24 EST 2011
Fixed argus.conf parsing bug, when using `hostname`.
Added some additional logging and packet dumping if the timestamps are wayyyy
out of the wayyyy.  Did a bit more cleanup on PSIZE DSR type reporting, due
to error messages when using PF_RING.

Thu Dec  1 17:32:30 EST 2011
Updated ifdef's for available pcap functions
Added back support for pcap_dispatch() to support older version of Bivio.

Mon Nov 14 14:29:08 EST 2011
Fixed problem with fragment tracking and reporting
Added stub for process Juniper TTP protocol

Mon Sep 19 10:01:48 EDT 2011
Modified output global status timer management to deal with sleeping laptops.

Fri Jun 30 10:22:43 EDT 2011
Fixed pid file code to create correct pid files.

Mon Jun 20 15:36:17 EDT 2011
Added 4 byte string and fixed int as probe id.

Mon May 16 18:44:22 EDT 2011
Added v1, 5, and 6 to this list of cisco formats.

Mon May  9 23:02:38 EDT 2011
Added parsing cisco flow records from packet payload (currently support v7).

Mon Apr 18 19:38:02 EDT 2011
Fix for ArgusMallocListRecord when we've run out of memory

Fri Feb 25 13:34:59 EST 2011
Added timeout variables to argus.conf, and removed the init timeout value.

Fri Feb 25 11:08:49 EST 2011
Fixes for Argus events, DAG interface support and configure.ac
Ready to go

Tue Oct 12 14:58:13 EDT 2010
Tuning for timeout queues

Mon Oct 11 15:53:47 EDT 2010
Updated for changes in argus-3.x.3.24
Merged in argus-3.0.3.17 into argus-3.x.3.

Thu Sep 30 19:29:40 EDT 2010
An amazing set of bugs relating to List records.
Should make argus much better under load

Mon Aug 16 14:36:54 EDT 2010
Reimplemenation of keystroke algorithm

Tue Aug 10 23:54:21 EDT 2010
Rework autoconf and autoheader support

Tue Aug 10 01:33:30 EDT 2010
Mods from John's email and added debuging

Thu Jul 29 14:04:51 EDT 2010
Finished first round implementation for keystroke algorithm

Sat Jul 17 17:33:42 EDT 2010
Fix to get localnet and netmask back into management record

Tue Jul  6 18:12:57 EDT 2010
Fix for fragmentation support

Wed Jun 23 14:20:18 EDT 2010
Fix for jitter value reporting.

Thu Jun 10 11:02:57 EDT 2010
Fix for "bond" interface support in ArgusSource.

Wed Jun  2 15:31:54 EDT 2010
Added multiple bind address support
Added udt packet stream processing
Added control plane capture support using -C
Fix for arp flow record reporting (missing answer)
Correct AI_NUMERICSERV change to AI_NUMERICHOST

Mon May 10 18:37:42 EDT 2010
Finishing threaded ArgusSource strategy.
Added back ArgusOutput udp strategy.
Updated argus.conf.5 man page.
Updated argus.8 man page.
Fixed gentoo configure support.

Tue May  4 10:31:14 EDT 2010
Added back threaded ArgusSource strategy.

Tue May  4 00:30:47 EDT 2010
Added gentoo bug reports for configure and compile
Added better syslog support.

Fri Apr 23 11:10:59 EDT 2010
Added ISIS flow generation for tony.

Added backtrace for argus exiting, and added a few checks for pcap_dispatch()
returning an ignorable error.

Thu Apr 22 11:38:09 EDT 2010
Mods to fix TCP loss reporting error due to incomplete implementation
of Mar 8 fixes for modifing the packet.
Mod to fix TCP state machine when FIN alone is sent from dst.

Mon Apr 12 22:45:54 EDT 2010
Mods to fix Teredo parsing error

Mon Mar  8 23:40:34 EST 2010
Mods to not modify the packet itself while processing.

Tue Feb 16 23:16:53 EST 2010
Modify teredo support to be turned on via argus.conf
Indicate teredo tunnel encapsulation by default.

Wed Feb 10 17:41:58 EST 2010
Added Teredo tunnel parsing.  Exclude udp domain as candidate.

Mon Feb  8 21:46:32 EST 2010
Starting argus-3.0.3
Added argus events data generation
Fixed apple libpcap performance issues.
Updated argusarchive with Peter's version (thanks Peter!!!)

Changes from Argus-3.0.0
Thu Oct 15 12:15:17 EDT 2009
Mod to support Bivio7000 default device processing.

Thu Sep 24 01:23:02 EDT 2009
Fixes for ipid roll-over and TCP loss detection

Thu Aug 13 00:40:12 EDT 2009
Fixes for GRE flow reporting when the upper protocol is incomplete
Fixes for over counting some fragments.

Tue Aug 11 01:05:23 EDT 2009
Mods to fix retrans reporting for tcp when packets are out of order

Sun Aug  9 19:32:27 EDT 2009
Mods to fix poor header parsing logic in pppoe protocols.

Tue Jul 28 10:57:18 EDT 2009
Mods to support older libpcap libraries, aka Bivio 7500
Mods to fix loss reporting for ESP flows.

Tue Jul  7 14:02:35 EDT 2009
OK, seems like a year has passed, and here we are for argus-3.0.2.
The changes include:
   Syslog support fixes
   Modifications to rpm argus.spec file.
   Modificiations to support Bivio 7500.
   Fixes for arp processing.
   Fixes for icmpv6 processing.
   Restructure timeout queue processing.
   Attempted fixes for dealing with system time bug.
   Fixes for Radiotap header parsing.
   Mods to ArgusProcessEtherHdr logic for PPP processing.
   Reimplement ARGUS_SRC_MULTIPATH;
   Fix packet size reporting error.

   Set ARGUS_INITTIMEOUT from 0 to 5 seconds.
   Restructure Listen strategy.
   Minor changes to coping of timestruct from libpcap to us.
   Better handling of pcap fileno.
   Correct MAX_SNAPLEN byte length to 65535
   Fix TCP flags reporting
   Fix TCP Reinitialization on RST.
   Corrections for long long variables by using LL in constants.
   Restructured source id porsing.

argus/ArgusArp.c
/*
 * Argus Software.  Argus files - Arp Procession
 * Copyright (c) 2000-2012 QoSient, LLC
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

/* 
 * $Id: //depot/argus/argus/argus/ArgusArp.c#27 $
 * $DateTime: 2012/01/03 19:15:19 $
 * $Change: 2274 $
 */

#ifdef HAVE_CONFIG_H
#include "argus_config.h"
#endif

#if !defined(ArgusArp)
#define ArgusArp
#endif

#include <stdio.h>
#include <argus_compat.h>
#include <ArgusModeler.h>

#if !defined(__OpenBSD__)
#include <net/if_arp.h>
#endif

#include <string.h>
#include <errno.h>

#define argus_ar_sha(ap)   (((const u_char *)((ap)+1))+0)
#define argus_ar_spa(ap)   (((const u_char *)((ap)+1))+  (ap)->ar_hln)
#define argus_ar_tha(ap)   (((const u_char *)((ap)+1))+  (ap)->ar_hln+(ap)->ar_pln)
#define argus_ar_tpa(ap)   (((const u_char *)((ap)+1))+2*(ap)->ar_hln+(ap)->ar_pln)

#define ARP_HDRLEN      8

#define HRD(ap) EXTRACT_16BITS(&(ap)->ar_hrd)
#define HLN(ap) ((ap)->ar_hln)
#define PLN(ap) ((ap)->ar_pln)
#define OP(ap)  EXTRACT_16BITS(&(ap)->ar_op)
#define PRO(ap) EXTRACT_16BITS(&(ap)->ar_pro)

#if defined(SHA)
#undef SHA
#endif
#define SHA(ap) (argus_ar_sha(ap))

#if defined(SPA)
#undef SPA
#endif
#define SPA(ap) (argus_ar_spa(ap))

#if defined(THA)
#undef THA
#endif
#define THA(ap) (argus_ar_tha(ap))

#if defined(TPA)
#undef TPA
#endif
#define TPA(ap) (argus_ar_tpa(ap))

#if !defined(REVARP_REQUEST)
#define REVARP_REQUEST          3
#endif
#if !defined(REVARP_REPLY)
#define REVARP_REPLY            4
#endif

void ArgusUpdateArpState (struct ArgusModelerStruct *, struct ArgusFlowStruct *, unsigned char *);

/*
struct ArgusHAddr {
   union {
      unsigned char ethernet[6];
      unsigned char ib[32];
      unsigned char ieee1394[16];
      unsigned char framerelay[4];
      unsigned char tokenring[6];
      unsigned char arcnet[1];
      unsigned char fiberchannel[12];
      unsigned char atm[20];
   } haddr_un;
};

struct ArgusArpFlow {
   unsigned short    hrd;
   unsigned short    pro;
   unsigned char     hln; 
   unsigned char     pln;
   unsigned short    op;
   unsigned int      arp_spa;
   unsigned int      arp_tpa;
   struct ArgusHAddr haddr;
};
 
struct ArgusRarpFlow {
   unsigned short    hrd;
   unsigned short    pro;
   unsigned char     hln;
   unsigned char     pln;
   unsigned short    op;
   unsigned int      arp_tpa;
   struct ArgusHAddr shaddr;
   struct ArgusHAddr dhaddr;
};
*/

#define ARP_ETHERNET	1
#define ARP_IEEE802	6
#define ARP_ARCNET  	7
#define ARP_FRELAY	15
#define ARP_ATM 	19
#define ARP_STRIP	23
#define ARP_IEEE1394	24
#define ARP_INFINIBAND	32

struct ArgusSystemFlow *
ArgusCreateArpFlow (struct ArgusModelerStruct *model, struct ether_header *ep) 
{
   struct arphdr *ahdr = (struct arphdr *)model->ArgusThisUpHdr;
   struct ArgusSystemFlow *retn = NULL;
   unsigned int arp_tpa, arp_spa;

   if (STRUCTCAPTURED(model, *ahdr)) {
      retn = model->ArgusThisFlow;

      retn->hdr.type              = ARGUS_FLOW_DSR;
      retn->hdr.subtype           = ARGUS_FLOW_ARP;

      switch (OP(ahdr)) {
         case ARPOP_REQUEST: {
            retn->hdr.argus_dsrvl8.len  = sizeof(struct ArgusArpFlow)/4 + 1;
            retn->hdr.argus_dsrvl8.qual = ARGUS_TYPE_ARP;

            if (PLN(ahdr) == sizeof(arp_spa)) {
               bcopy (SPA(ahdr), &arp_spa, sizeof(arp_spa));
               bcopy (TPA(ahdr), &arp_tpa, sizeof(arp_tpa));

#ifdef _LITTLE_ENDIAN
               arp_spa = ntohl(arp_spa);
               arp_tpa = ntohl(arp_tpa);
#endif
               if (arp_spa > arp_tpa)
                  model->state |= ARGUS_DIRECTION;

               retn->arp_flow.hrd = HRD(ahdr);
               retn->arp_flow.pro = PRO(ahdr);
               retn->arp_flow.hln = HLN(ahdr);
               retn->arp_flow.pln = PLN(ahdr);
               retn->arp_flow.op  =  OP(ahdr);

               retn->arp_flow.arp_tpa = arp_tpa;
               retn->arp_flow.arp_spa = arp_spa;

               bcopy (SHA(ahdr), (char *)&retn->arp_flow.haddr, HLN(ahdr));
            }
            break;
         }
   
         case ARPOP_REPLY: {
            retn->hdr.argus_dsrvl8.len  = sizeof(struct ArgusArpFlow)/4 + 1;
            retn->hdr.argus_dsrvl8.qual = ARGUS_TYPE_ARP;

            if (PLN(ahdr) == sizeof(arp_spa)) {
               bcopy (SPA(ahdr), &arp_spa, sizeof(arp_spa));
               bcopy (TPA(ahdr), &arp_tpa, sizeof(arp_tpa));

#ifdef _LITTLE_ENDIAN
               arp_spa = ntohl(arp_spa);
               arp_tpa = ntohl(arp_tpa);
#endif
               if (arp_spa > arp_tpa)
                  model->state |= ARGUS_DIRECTION;

               retn->arp_flow.hrd     = HRD(ahdr);
               retn->arp_flow.pro     = PRO(ahdr);
               retn->arp_flow.hln     = HLN(ahdr);
               retn->arp_flow.pln     = PLN(ahdr);
               retn->arp_flow.op      = ARPOP_REQUEST;

               retn->arp_flow.arp_tpa = arp_spa;
               retn->arp_flow.arp_spa = arp_tpa;
   
               bcopy (THA(ahdr), (char *)&retn->arp_flow.haddr, HLN(ahdr));
            }
            break;
         }

         case REVARP_REQUEST: {
            retn->hdr.argus_dsrvl8.len  = sizeof(struct ArgusRarpFlow)/4 + 1;
            retn->hdr.argus_dsrvl8.qual = ARGUS_TYPE_RARP;

#ifdef _LITTLE_ENDIAN
            arp_tpa = ntohl(arp_tpa);
#endif
            retn->rarp_flow.hrd     = HRD(ahdr);
            retn->rarp_flow.pro     = PRO(ahdr);
            retn->rarp_flow.hln     = HLN(ahdr);
            retn->rarp_flow.pln     = PLN(ahdr);
            retn->rarp_flow.op      =  OP(ahdr);

            bcopy (THA(ahdr), &retn->rarp_flow.shaddr, HLN(ahdr));
            bcopy (SHA(ahdr), &retn->rarp_flow.dhaddr, HLN(ahdr));
            break;
         }

         case REVARP_REPLY: {
            retn->hdr.argus_dsrvl8.len  = sizeof(struct ArgusRarpFlow)/4 + 1;
            retn->hdr.argus_dsrvl8.qual = ARGUS_TYPE_RARP;
            bcopy (TPA(ahdr), &arp_tpa, sizeof(arp_tpa));

#ifdef _LITTLE_ENDIAN
            arp_tpa = ntohl(arp_tpa);
#endif
            retn->rarp_flow.hrd     = HRD(ahdr);
            retn->rarp_flow.pro     = PRO(ahdr);
            retn->rarp_flow.hln     = HLN(ahdr);
            retn->rarp_flow.pln     = PLN(ahdr);
            retn->rarp_flow.op      = REVARP_REQUEST;

            bcopy ((char *)&arp_tpa, &retn->rarp_flow.arp_tpa, sizeof(arp_tpa));
            bcopy (SHA(ahdr), &retn->rarp_flow.shaddr, HLN(ahdr));
            bcopy (THA(ahdr), &retn->rarp_flow.dhaddr, HLN(ahdr));
            break;
         }
      }
   }

#ifdef ARGUSDEBUG
   ArgusDebug (8, "ArgusCreateArpFlow (0x%x) returning %d\n", ep, retn);
#endif

   return (retn);
}

/*
struct  arphdr {
        u_short ar_hrd;
#define ARPHRD_ETHER    1
#define ARPHRD_IEEE802  6
#define ARPHRD_FRELAY   15
#define ARPHRD_IEEE1394 24
#define ARPHRD_IEEE1394_EUI64 27
        u_short ar_pro;
        u_char  ar_hln;
        u_char  ar_pln;
        u_short ar_op;
#define ARPOP_REQUEST   1
#define ARPOP_REPLY     2
#define ARPOP_REVREQUEST 3
#define ARPOP_REVREPLY  4
#define ARPOP_INVREQUEST 8
#define ARPOP_INVREPLY  9
#ifdef COMMENT_ONLY
        u_char  ar_sha[];
        u_char  ar_spa[];
        u_char  ar_tha[];
        u_char  ar_tpa[];
#endif
};
*/

void
ArgusUpdateArpState (struct ArgusModelerStruct *model, struct ArgusFlowStruct *flowstr, unsigned char *state)
{
   struct ArgusARPObject *arpobj = NULL;
   struct ether_arp *arp = NULL;
   struct arphdr *ahdr = NULL;

   if (model->ArgusThisEpHdr == NULL)
      return;

   arp = (struct ether_arp *)(model->ArgusThisEpHdr + 1);
   ahdr = &arp->ea_hdr;

   if (STRUCTCAPTURED(model, *arp)) {
      model->ArgusThisLength -= sizeof(*arp);
      model->ArgusSnapLength -= sizeof(*arp);
      model->ArgusThisUpHdr = (unsigned char *)(arp + 1);

      switch (OP(ahdr)) {
         case ARPOP_REQUEST:
            break;

         case ARPOP_REPLY: {
            struct ArgusNetworkStruct *net = (struct ArgusNetworkStruct *) flowstr->dsrs[ARGUS_NETWORK_INDEX];

            if (net == NULL) {
               net = (struct ArgusNetworkStruct *) &flowstr->canon.net;
               flowstr->dsrs[ARGUS_NETWORK_INDEX] = (struct ArgusDSRHeader *) net;
            }

            net->hdr.type              = ARGUS_NETWORK_DSR;
            net->hdr.subtype           = ARGUS_NETWORK_SUBTYPE_ARP;
            net->hdr.argus_dsrvl8.qual = 0;
            net->hdr.argus_dsrvl8.len  = ((sizeof(struct ArgusUDTObject) + 3)/4) + 1;

            arpobj = &net->net_union.arp;

            bcopy ((unsigned char *)SHA(ahdr), arpobj->respaddr, 6);
            break;
         }
      }
   }
}
